### PR TITLE
Add benchmark testing framework

### LINKED ISSUE #1365: Envoy Gateway performance at scale
*Description*:
This issue tracks the performance (Throughput, Latency) of Envoy Gateway control plane and data plane at scale (Service, xRoutes, Gateway, Client Connections) . 
  * The output of this issue should be a document that can be presented to the user to better understand what is the performance of Envoy Gateway by default and how can this be increased if there is an increase in scale.
  * Other issues might need to be created (based on testing) and linked to this issue 

[optional *Relevant Links*:]
Emissary: https://www.getambassador.io/docs/emissary/latest/topics/running/scaling
Contour: https://github.com/projectcontour/contour-perf / https://projectcontour.io/guides/resource-limits/
Istio: https://istio.io/v1.16/docs/ops/deployment/performance-and-scalability/

### LINKED ISSUE #2325: Add Performance test in CI
*Description*:

some ideas:

1. Investigate possibilities to integrate EG with KWOK for performance test.

cc @wzshiming

[optional *Relevant Links*:]

https://github.com/kubernetes-sigs/kwok

Part of https://github.com/envoyproxy/gateway/issues/1365

### PR DESCRIPTION
**Pull request by shawnh2**
**What this PR does / why we need it**:

Taking https://github.com/envoyproxy/gateway/pull/2578 forward, but using code to implement benchmark testing framework.

Similar to `ConformanceSuite` and `ConformanceTest`, I define a `BenchmarkSuite` and `BenchmarkTest`, each benchmark test case will be run as a k8s Job, so that:
- we can easily control our benchmark env like Pod's resources limit and request
- we can add more test cases in the future without any breaking changes

**Which issue(s) this PR fixes**:

Fix #1365, Fix #2325, Close #2578

### CODE REVIEW

**Comment by shawnh2** [comment]
For now, we have successfully run the benchmark test, for example, see logs for all the cases, the scale of each httproutes are for test purpose only, will update them to a bigger number after everything has done:
- https://github.com/envoyproxy/gateway/actions/runs/9525887143/job/26260593652?pr=3599#step:4:1111
- https://github.com/envoyproxy/gateway/actions/runs/9525887143/job/26260593652?pr=3599#step:4:1232
- https://github.com/envoyproxy/gateway/actions/runs/9525887143/job/26260593652?pr=3599#step:4:1374

**Comment by zirain** [comment]
where's the result?

**Comment by shawnh2** [comment]
> where's the result?

In the CI stdout, which way do we prefer to see the result ? Comments back in current thread ?

**Comment by zirain** [comment]
> > where's the result?
> 
> In the CI stdout, which way do we prefer to see the result ? Comments back in current thread ?

that's an option.

File: .github/workflows/benchmark.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+name: Benchmarking Tests at Scale
```
[the remark this answers is not in the document]
Comment (arkodg) [reply .github/workflows/benchmark.yaml:1]: id vote to only make it run on push to `main` and `release/v*`

lets raise a follow up issue to support running on PRs automatically (if it doesn't increase CI time) or using `/benchmark`

File: test/benchmark/config/gatewayclass.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+metadata:
+  name: proxy-config
+  namespace: envoy-gateway-system
+spec:
+  provider:
+    type: Kubernetes
+    kubernetes:
+      envoyDeployment:
+        container:
+          volumeMounts:
```
Comment (arkodg) [inline test/benchmark/config/gatewayclass.yaml:24]: is the ext proc config required ?

File: test/benchmark/config/gatewayclass.yaml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            - mountPath: /var/run/ext-proc
+              name: socket-dir
+        pod:
```
Comment (arkodg) [inline test/benchmark/config/gatewayclass.yaml:27]: we should probably set high cpu and men limits for this test

File: test/benchmark/tests/scale_httproutes.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		// Setup and create a gateway.
+		gatewayNN := types.NamespacedName{Name: "benchmark", Namespace: ns}
+		gateway := bSuite.GatewayTemplate.DeepCopy()
+		gateway.SetName(gatewayNN.Name)
+		err = bSuite.CreateResource(ctx, gateway)
+		require.NoError(t, err)
+
+		// Setup and scale httproutes.
+		httpRouteName := "benchmark-route-%d"
+		httpRouteScales := []uint16{10, 50, 100, 300, 500}
```
Comment (arkodg) [inline test/benchmark/tests/scale_httproutes.go:50]: lts also tracking scaling down, so lets also add `10, 50, 100, 300, 500, 300, 100, 50`

**Comment by arkodg** [comment]
thanks for building out this benchmarking suite !

suggest configuring envoy proxy and envoy gateway memory and cpu limits before starting the test, and also make it a top level input arg
I like how this is a suite of individual tests, each test changing some parameter, the first one modifies `HTTPRoute`.
A suggestion is to graph the values of the below attributes with changes in the input param (e.g. HTTPRoute)
* Latency (P50, P90 and P99)
* Error Rate (non 200)
* Throughput
* EG Memory usage
* EG CPU usage
* EnvoyProxy Memory usage
* EnvoyProxy CPU usage
for each test within `Benchmark`

**Comment by arkodg** [comment]
we are already using fortio in our repo today https://github.com/envoyproxy/gateway/blob/2c602d518a5be35e9b961092278c420ebc16e8b8/test/e2e/tests/utils.go#L154
are they any benefits on using `nighthawk` here ?
cc @guydc

File: .github/workflows/benchmark.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (shawnh2) [reply .github/workflows/benchmark.yaml:1]: sounds good!

**Comment by shawnh2** [comment]
> A suggestion is to graph the values of the below attributes with changes in the input param (e.g. HTTPRoute)
> 

Cool! These values can be easily retrieved from CP metrics and graphed out.

File: test/benchmark/config/gatewayclass.yaml (hunk 2 above)
Comment (shawnh2) [reply test/benchmark/config/gatewayclass.yaml:24]: no. this content is directly copied from https://github.com/envoyproxy/gateway/blob/main/test/config/gatewayclass.yaml w/o any modifications.

I will check this file again, add some resource limits and remove useless configs.

File: test/benchmark/config/gatewayclass.yaml (hunk 3 above)
Comment (shawnh2) [reply test/benchmark/config/gatewayclass.yaml:27]: got it!

File: test/benchmark/config/gatewayclass.yaml (hunk 2 above)
Comment (shawnh2) [reply test/benchmark/config/gatewayclass.yaml:24]: removed

File: .github/workflows/benchmark.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+on:
+  pull_request:
```
Comment (shawnh2) [inline .github/workflows/benchmark.yaml:3]: change it to `push` once this PR is good to go

File: test/benchmark/config/nighthawk-test-server.yaml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    matchLabels:
+      app: nighthawk-test-server
+  template:
+    metadata:
+      labels:
+        app: nighthawk-test-server
+    spec:
+      serviceAccountName: default
+      containers:
+        - name: nighthawk-server
```
Comment (shawnh2) [inline test/benchmark/config/nighthawk-test-server.yaml:19]: can replace this test server with a much simpler one, like echo server. can be done as a follow-up PR.

File: test/benchmark/tests/scale_httproutes.go (hunk 4 above)
Comment (shawnh2) [reply test/benchmark/tests/scale_httproutes.go]: added the scale-down test case and support

**Comment by shawnh2** [comment]
Looks like the 1024MiB memory for EG is not enough when we are trying to scale httproutes to 500.

The memory changes:

| HTTPRoutes scale | Memory consumption |
| - | - |
| 10 | [78.7MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:1406) |
| 50 | [97.3MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:2103) |
| 100 | [184.3MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:2921) |
| 300 | [744.6MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:4489) |
| 500 | > 1024MiB |

So increase memory to 2048MiB now.

**Comment by Xunzhuo** [comment]
> Looks like the 1024MiB memory for EG is not enough when we are trying to scale httproutes to 500.
> 
> The memory changes:
> 
> HTTPRoutes scale	Memory consumption
> 10	[78.7MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:1406)
> 50	[97.3MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:2103)
> 100	[184.3MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:2921)
> 300	[744.6MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:4489)
> 500	> 1024MiB
> Should we increase the memory to 2048MiB or fixed the max scale of httproutes to 300 ?

@shawnh2  1 gwc + 1 gw + 500 httproutes will cost eg over 1 GB memory ?

File: .github/workflows/benchmark.yaml (hunk 5 above)
Comment (Xunzhuo) [reply .github/workflows/benchmark.yaml:3]: Not sure if we need to run this in every push, or on schedule

**Comment by Xunzhuo** [comment]
image

when logging condition met, I think we need more complete info for that, to clearly know what what resource condition met.

**Comment by shawnh2** [comment]
> @shawnh2 1 gwc + 1 gw + 500 httproutes will cost eg over 1 GB memory ?

Yes, according to `process_resident_memory_bytes` metrics collected from CP, that will cost approximately [1.9GiB](https://github.com/envoyproxy/gateway/actions/runs/9580041576/job/26413815253?pr=3599#step:4:6456) of memory.

**Comment by shawnh2** [comment]
> image
> when logging condition met, I think we need more complete info for that, to clearly know what what resource condition met.

Currently, the `BenchmarkSuite` reuse the helper functions from upstream GW-API. I guess we can add more log info in our test code, without changing any code from upstream.

**Comment by arkodg** [comment]
> Looks like the 1024MiB memory for EG is not enough when we are trying to scale httproutes to 500.
> 
> The memory changes:
> 
> HTTPRoutes scale	Memory consumption
> 10	[~ 70MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:1406)
> 50	[~ 100MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:2103)
> 100	[~ 200MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:2921)
> 300	[~ 700MiB](https://github.com/envoyproxy/gateway/actions/runs/9565607105/job/26368897505?pr=3599#step:4:4489)
> 500	~ 1800MiB
> Should we increase the memory to 2048MiB or fixed the max scale of httproutes to 300 ?

lets increase to 2GB and raise a follow up to triage why the increase is not linear

**Comment by Xunzhuo** [comment]
@shawnh2 Can we have something like benchmark report ?

**Comment by shawnh2** [comment]
> @shawnh2 Can we have something like benchmark report ?

Sure do, here is a example report on my test branch https://github.com/shawnh2/gateway/pull/11#issuecomment-2182133520. I will update it here.

**Comment by Xunzhuo** [comment]
But I think we cannot have that on pull_request, the workflow does not have that privilege

**Comment by shawnh2** [comment]
You mean add this report like a doc ?

**Comment by Xunzhuo** [comment]
Nope, I agreed with that this should be along with a PR, but I think the pull_request do not have the right to output the report. You tested that in your own report, have you tested it here ?

**Comment by zhaohuabing** [comment]
> Yes, it does not have privilege to preview the report as a comment, I shall upload a report along with this PR.

The report can be posted to a PR as a comment with `pull_request_target` event:

```
name: Benchmarking Tests at Scale
on:
  pull_request_target:
    branches:
      - "main"
      - "release/v*"
  workflow_dispatch:
    inputs:
.....    
```

But I'm not sure if there're any security concern with this approach.

File: .github/workflows/benchmark.yaml (hunk 5 above)
Comment (shawnh2) [reply .github/workflows/benchmark.yaml:3]: like suggested https://github.com/envoyproxy/gateway/pull/3599#discussion_r1643460287, we can run this with `/benchmark` command.

**Comment by shawnh2** [comment]
> > Yes, it does not have privilege to preview the report as a comment, I shall upload a report along with this PR.
> 
> The report can be posted to a PR as a comment with `pull_request_target` event:

Still not working, the error shows `Resource not accessible by integration`, stick to upload report along with this PR.

File: test/benchmark/config/httproute.yaml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+apiVersion: gateway.networking.k8s.io/v1
+kind: HTTPRoute
+metadata:
+  name: "{HTTPROUTE_NAME}"
+  namespace: benchmark-test
+spec:
+  parentRefs:
+    - name: "{REF_GATEWAY_NAME}"
+  hostnames:
+    - "www.benchmark.com"
```
Comment (arkodg) [inline test/benchmark/config/httproute.yaml:10]: can we also template out the hostname for this test so each HTTPRoute gets a unique hostname
else these routes wont reach `Programmed` and will bloat `Status`

File: test/benchmark/benchmark_report.md [hunk 8]
```diff
@@ -0,0 +1,925 @@
+# Benchmark Report
+
+Benchmark test settings:
+
+|RPS |Connections|Duration (Seconds)|CPU Limits (m)|Memory Limits (MiB)|
+|-   |-          |-                 |-             |-                  |
+|1000|100        |90                |1000          |2048               |
```
Comment (arkodg) [inline test/benchmark/benchmark_report.md:7]: does each step (e.g. HTTPRoute = 50 or HTTPRoute = 300) run for 90s ?

File: test/benchmark/benchmark_report.md [hunk 9]
```diff
@@ -0,0 +1,925 @@
+# Benchmark Report
```
Comment (arkodg) [inline test/benchmark/benchmark_report.md:1]: ❤️  thanks for generating this !

**Review by arkodg — APPROVED** [review]
LGTM thanks for adding this framework and the report !

**Comment by arkodg** [comment]
> > are they any benefits on using nighthawk here ?
> 
> Nighthawk:
> 
> * Official benchmarking tool for envoy: makes performance-related discussions with upstream envoy easier, as it help rule-out the loadgen tech as a culprit for issues.
> * Uses the same configuration as envoy proxy (easier for us to define client and server behavior, supports very advanced settings)
> 
> Fortio:
> 
> * Easier to interact with in golang (download, compile, call)
> * Very popular and maintained performance tool in the envoy ecosystem
> 
> Istio supports execution of their dataplane benchmark tests with either Fortio or Nighthawk, the latter being the most recent addition: [istio/istio#21161](https://github.com/istio/istio/issues/21161). Istio load tests are currently executed with fortio.
> 
> Overall, I'm +1 for using nighthawk as the data plane benchmark tool. The current use of fortio in our project is mostly for easy execution of load during e2e tests that require it (retries, shutdown, upgrade, ... ). I believe that we can implement our own simple load generator based on net/http and remove the fortio dependency in the future.

non blocking comment, my preference would be to run the client outside the k8s cluster, which is easier with fortio as a golang lib rather than running a containerized nighthawk client

File: .github/workflows/benchmark.yaml (hunk 5 above)
Comment (AliceProxy) [reply .github/workflows/benchmark.yaml:3]: I think it is good to run it on pull/push. In general, I like all major testing/linting/etc. CI suites to run on every push even if there is no PR. Makes it easy to get your branches in order without having a PR that goes through a bunch of edits to get things working. I dislike the idea of only ever running it when users comment `/benchmark`. The general idea should be for CI to alert us when incoming changes degrade (or improve) performance.

**Review by AliceProxy — APPROVED** [review]

**Review by guydc — APPROVED** [review]
LGTM! Looking forward to seeing this run in CI!

**Comment by arkodg** [comment]
there's a lint error
```
Error: ./test/benchmark/benchmark_report.md:814: ocurred ==> occurred
```

File: .github/workflows/benchmark.yaml [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    branches:
+      - "main"
+      - "release/v*"
+  workflow_dispatch:
+    inputs:
+      rps:
+        description: "The target requests-per-second rate. Default: 1000"
+        default: '1000'
```
[the remark this answers is not in the document]
Comment (shawnh2) [reply .github/workflows/benchmark.yaml:11]: sure

File: test/benchmark/benchmark_report.md (hunk 8 above)
Comment (shawnh2) [reply test/benchmark/benchmark_report.md:7]: yes

File: test/benchmark/config/httproute.yaml (hunk 7 above)
Comment (shawnh2) [reply test/benchmark/config/httproute.yaml:10]: make sense, do it as a follow-up

**Comment by shawnh2** [comment]
> there's a lint error
> 
> ```
> Error: ./test/benchmark/benchmark_report.md:814: ocurred ==> occurred
> ```

Seems like a typo from nighthawk

**Review by arkodg — APPROVED** [review]

**Review by guydc — APPROVED** [review]

**Pull request merged**